Introduce Parathreads (runtime)

gavofyork opened the pull request:
~TODO~
- [x] Upgrade to latest Substrate + https://github.com/paritytech/substrate/pull/3213 (using that branch and premerged with #340 )
- [x] Make build
- [x] Fix tests
- [x] Add new tests for parathread and altered parachain functionality.
- [x] Introduce new modules and linkage to the runtime libs.rs
- [x] Refactor `TypeId` so it doesn't pull in `sr_io` and the memory management stuff.
- [x] Check that the accumulator logic (which uses the chain's state) isn't an issue for the tx queue otherwise we'll construct invalid blocks when there's more than the `ThreadCount` number of selection transactions waiting

gavofyork commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,355 @@
+				ensure!(actual == head_data_hash, DispatchError::Stale);
+
+				// updated the selected threads.
+				selected_threads.insert(pos, (id, collator));
```
@tomusdrw don't know how well this will work with the current txqueue model... it assumes that we retain these changes in the candidate block's state *as we populate* our candidate block with txs from the tx queue.

tomusdrw replied:
I don't fully understand how it relates to the pool, but if you call `validate` not only during `validate_transaction` but also during `apply_extrinsic` the changes will indeed be retained.
However when we import extrinsics to the pool they will runt `validate_transaction` on a pristine state of latest block and the changes are obviously not persisted.

Is that what you had in mind or did I miss your point completely?

gavofyork replied:
it should work ok like that, i think.

but generally, these transactions are a little more sophisticated than usual. there are several instances when they can become invalid:

1. the head-data-hash given in them is not (or no longer) the hash of the parathread's head-data. this can happen because:
  a. it was old when we received it (should never be imported into pool)
  b. it was an alternative fork (should never be imported into pool)
  c. the parathread has progressed since we received it (should be thrown out as soon as this happens).
2. two of them are for the same parathread with the same head-data-hash. in this case the one with the lower fee/priority should be discarded as the other will always supersede it.
3. some number (`registry::ThreadCount`) of parathread progressions have already occurred in this block. in this case, the transactions should remain in the pool as long as their head-data-hash has not become outdated.

as i understand it, the pool should import them as long as the HDH is current, it will correctly construct blocks with them properly prioritised (this just uses the fees priority) and it will avoid putting in more than it should into a block.

to make it proactively discard duplicates, we presumably need to ensure that their `provides` includes the same tag, which should fulfil the needs for (2).

but in order to get (1c) working, we'd need to essentially re-check the validity after each block is imported. does the queue do this already?

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		/// The number of threads to schedule per block.
+		ThreadCount: u32;
+
+		/// Current set of threads the next block; ordered by para ID. There can be no duplicates
```
```suggestion
		/// Current set of threads in the next block; ordered by parachain ID. There can be no duplicates
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		ThreadCount: u32;
+
+		/// Current set of threads the next block; ordered by para ID. There can be no duplicates
+		/// of para ID in the list.
```
```suggestion
		/// of parachain ID in the list.
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+				register_para(id, info, code, initial_head_data)
+		}
+
+		/// Deregister a parachain with given id
```
```suggestion
		/// Deregister a given parachain ID
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+			let id = crate::parachains::ensure_parachain(origin)?;
+			let debtor = T::Lookup::lookup(debtor)?;
+
+			let info = Paras::get(id).ok_or("invalid id")?;
```
```suggestion
			let info = Paras::get(id).ok_or("invalid parachain id")?;
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		fn swap(origin, #[compact] other: ParaId) {
+			let id = crate::parachains::ensure_parachain(origin)?;
+
+			let info = Paras::get(id).ok_or("invalid id")?;
```
```suggestion
			let info = Paras::get(id).ok_or("invalid parachain id")?;
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+
+impl<T: Trait> Module<T> {
+	pub fn ensure_thread_id(id: ParaId) -> rstd::result::Result<ParaInfo, &'static str> {
+		let info = Paras::get(id).ok_or("invalid id")?;
```
```suggestion
		let info = Paras::get(id).ok_or("invalid parachain id")?;
```

ltfschoen commented on runtime/src/slots.rs:
```diff
@@ -122,9 +125,9 @@ decl_storage! {
 		/// The number of auctions that been started so far.
```
```suggestion
		/// The number of auctions started so far.
```

ltfschoen commented on runtime/src/slots.rs:
```diff
@@ -785,7 +816,6 @@ impl<T: Trait> Module<T> {
 }
 
-
 /// tests for this module
```
```suggestion
```

ltfschoen commented on parachain/src/lib.rs:
```diff
@@ -95,6 +94,12 @@ pub struct ValidationResult {
+/// Type for determining the active set of parachains.
+pub trait ActiveThreads {
+	/// Return the current ordered set of `Id`s of active parathreads.
+	fn active_threads() -> Vec<Id>;
```
```suggestion
	fn active_threads() -> Parathreads;
```

suggest adding type alias `pub type Parathreads = Vec<Id>;`

ltfschoen commented on parachain/src/lib.rs:
```diff
@@ -95,6 +94,12 @@ pub struct ValidationResult {
 #[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize, Debug))]
 pub struct Id(u32);
```
suggest move and import this with something like the following as has been done elsewhere in this PR:
```
use primitives::parachain::{
  Id as ThreadId
}
```

ltfschoen commented on primitives/src/parachain.rs:
```diff
@@ -53,6 +54,56 @@ pub type ValidatorIndex = u32;
+	///
+	/// NOTE: The initial implementation simply concatenates the (ordered) set of (permanent)
+	/// parachain IDs with the (unordered) set of parathread IDs selected for this block.
+	fn active_paras() -> Vec<(Id, Option<CollatorId>)>;
```
```suggestion
	fn active_paras() -> Parachains;
```

suggest adding the type alias `pub type Parachains = Vec<(Id, Option<CollatorId>)>;` as well

ltfschoen commented on primitives/src/parachain.rs:
```diff
@@ -361,7 +412,7 @@ substrate_client::decl_runtime_apis! {
 		fn duty_roster() -> DutyRoster;
 		/// Get the currently active parachains.
-		fn active_parachains() -> Vec<Id>;
+		fn active_parachains() -> Vec<(Id, Option<CollatorId>)>;
```
```suggestion
		fn active_parachains() -> Parachains;
```

same as per previous comment https://github.com/paritytech/polkadot/pull/341/files#r308987015

ltfschoen commented on runtime/src/lib.rs:
```diff
@@ -483,8 +505,8 @@ impl_runtime_apis! {
 		}
-		fn active_parachains() -> Vec<parachain::Id> {
-			Parachains::active_parachains()
+		fn active_parachains() -> Vec<(parachain::Id, Option<parachain::CollatorId>)> {
```
```suggestion
		fn active_parachains() -> Parachains {
```

as per comment https://github.com/paritytech/polkadot/pull/341/files#r308987015 we could reuse a type alias to simplify

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		/// Register a parachain with given code.
+		/// Fails if given ID is already used.
+		#[weight = SimpleDispatchInfo::FixedOperational(5_000_000)]
+		pub fn register_para(origin,
```
The module's [`register_para` public function](https://github.com/paritytech/polkadot/pull/341/files#diff-6ceafc4b4dd2600a9e5caebc167c0cecR181) here has `info` as the last argument, whereas the `register_para` traits https://github.com/paritytech/polkadot/pull/341/files#diff-6ceafc4b4dd2600a9e5caebc167c0cecR181 and https://github.com/paritytech/polkadot/pull/341/files#diff-6ceafc4b4dd2600a9e5caebc167c0cecR65 have it as the 2nd argument. Would it add any value to have the argument ordering consistent?

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		///
+		/// Must be sent from a Signed origin that is able to have ParathreadDeposit reserved.
+		/// `code` and `initial_head_data` are used to initialize the parathread's state.
+		fn register_parathead(origin,
```
```suggestion
		fn register_parathread(origin,
```

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+	fn register_para(
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
```
A few of the functions have arguments like `(..., code: Vec<u8>, initial_head_data: Vec<u8>, ...)`.

Could we declare and then import as necessary some type aliases like:
```
pub type CodeHash = T::Hash;
pub type InitialHeadData = T::Hash;
```

And then change all those function arguments to `(..., code: CodeHash, initial_head_data: InitialHeadData, ...)`.

ltfschoen commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,381 @@
+		Paras get(paras): map ParaId => Option<ParaInfo>;
+	}
+	add_extra_genesis {
+		config(parachains): Vec<(ParaId, Vec<u8>, Vec<u8>)>;
```
Could we use type aliases (as mentioned in previous comment https://github.com/paritytech/polkadot/pull/341/files#r308993502) to change `Vec<u8>, Vec<u8>` to something like `CodeHash, InitialHeadData` (if my assumption of what they represent is correct)?

rphmeier commented on primitives/src/parachain.rs:
```diff
@@ -53,6 +53,56 @@ pub type ValidatorIndex = u32;
+	/// May only be called on a state that `can_swap` has previously returned `Ok` for: if this is
+	/// not the case, the result is undefined. May only return an error if `can_swap` also returns
+	/// an error.
+	fn do_swap(one: Id, other: Id) -> Result<(), &'static str>;
```
should be `on_swap` to match the trait name

joepetrowski commented on runtime/src/slots.rs:
```diff
@@ -155,22 +158,46 @@ decl_storage! {
+}
+
+impl<T: Trait> OnSwap for Module<T> {
+	fn can_swap(one: ParaId, other: ParaId) -> Result<(), &'static str> {
```
Returning `bool` on a `can_*` function would be more idiomatic, no?
```suggestion
	fn can_swap(one: ParaId, other: ParaId) -> bool {
		!<Onboarding<T>>::exists(one) && !<Onboarding<T>>::exists(other)
	}
```

shawntabrizi commented:
FYI I talked to @rphmeier who suggested we put this PR on ice until networking changes make their way into Polkadot, and we can get the base level Cumulus/Parachain interaction working.

Would be happy to pick back up when appropriate

gavofyork commented:
i'm not sure why @rphmeier would have said that, though it's true that parathreads will place somewhat more strenuous requirements on networking in full operation, it is most certainly not a blocker to an initial (and correct) prototype and/or mvp of parathreads, nor does it prevent it from being developed.

in the future, please avoid having private chats and then acting on "suggestions" without my approval. @rphmeier in the future, if you want to halt work on one of my PRs, come see me.

gavofyork commented:
@shawntabrizi a couple of changes we'll want to get in:

- We should plan which threads will be active one block in advance; this gives networking a chance (one full block's worth of time) to preconnect PVs with the winning collators in order to distribute the candidate blocks.
- We should consider allowing threads which missed a slot they paid for an extra one or two blocks' grace period. One way to do this would be to have one slot of spare capacity in each block and a queue of missed slots. Collators that miss their slot get placed in the missed slot queue. Any time a collator of a chain wins, then if there's an entry for that chain in missed queue, it's removed. The PV for extra slot of each block is just the head of the missed queue. If the entry in the missed queue misses their slot, then they're placed on the back of the queue (probably only for a limited number of times before they're discarded entirely).

CC @AlistairStewart

gavofyork commented:
That's the first item coded, though will need tests.

gavofyork commented:
Second one should wait until Al has weighed in.

rphmeier commented:
My intention was not to halt any work on the PR - only to delay a merge upon completion until the story around networking was more fleshed out.

shawntabrizi commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1297 @@
+						rstd::mem::swap(i, j)
+					)
+				);
+
```
```suggestion
				<Debtors<T>>::mutate(id, |i|
					<Debtors<T>>::mutate(other, |j|
						rstd::mem::swap(i, j)
					)
				);
```

This suggestion should fix the test I introduced which is currently failing.

Basically, if we swap a parathread with a parachain, update the paraid associated with a debtor such that if the _new_ parathread gets deregistered, the original debtor gets their funds returned.

Fund returning with a parachain is still managed by the `Offboarding` storage, and should be manually updated.

gavofyork replied:
bit weird, that the original debtor will get the funds back from a different parathread being exited, but probably nothing better that we can do...

shawntabrizi commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+			if PendingSwap::get(other) == Some(id) {
+				// actually do the swap.
+				T::OnSwap::ensure_can_swap(id, other)?;
+
```
```suggestion

// Remove intention to swap.
PendingSwap::remove(other);
```

This cleans up the "intention to swap" of a para which has successfully done a swap.

I updated a test to check for this, and this will make that test pass.

rphmeier replied:
right - without this there is a storage leak

shawntabrizi commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+const QUEUE_SIZE: usize = 2;
+
+/// The number of rotations that you will have as grace if you miss a block.
+const MAX_RETRIES: u32 = 3;
```
Any reason not to make these trait constants?

gavofyork replied:
Not really. Will do it.

shawntabrizi commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+				} else {
+					vec![]
+				};
+				if t.len() < QUEUE_SIZE {
```
```suggestion
				while t.len() < QUEUE_SIZE {
```

At worst, it is the same as `if`. At best, it will populate the empty queue stuff ensuring `SelectedThreads` always has `QUEUE_SIZE` length.

rphmeier commented on network/src/gossip.rs:
```diff
@@ -265,7 +265,7 @@ impl<F, P> ChainContext for (F, P) where
 		let active_parachains = api.active_parachains(&leaf_id)?;
 
-		for para_id in active_parachains {
+		for (para_id, _) in active_parachains {
```
```suggestion
        // TODO: https://github.com/paritytech/polkadot/issues/467
		for (para_id, _) in active_parachains {
```

rphmeier commented on primitives/src/parachain.rs:
```diff
@@ -78,6 +78,67 @@ pub type ValidatorPair = validator_app::Pair;
+}
+
+/// Type determining the active set of parachains in current block.
+pub trait ActiveParas {
```
```suggestion
pub trait ActiveParachains {
```

typically we avoid abbreviation (`use primitives::parachain::Id as ParaId` is the only exception, because it is so pervasive)

gavofyork replied:
I've tended to use "paras" to refer to both parachains and parathreads; in so far as the code that this PR touches, i've tried to reserve "parachains" to specifically those which are (semi-)permanent.

rphmeier commented on primitives/src/parachain.rs:
```diff
@@ -78,6 +78,67 @@ pub type ValidatorPair = validator_app::Pair;
+};
+
+/// Handler for when two parachains/parathreads get notionally swapped.
+pub trait OnSwap {
```
it seems that both functions `ensure_can_swap` and `on_swap` can be unified. This is similar to the Rust iterator trait vs. C++ iterators. In C++ you would usually have two methods: one for asking if the next item is available, and one for getting it. In Rust, they are combined.

gavofyork replied:
SRML dispatchables must first check that an operation will succeed in full before embarking on any mutating aspect of that operation. thus, a typical pattern for extensible operation components is a two part API check and enact. this is no different.

rphmeier commented on primitives/src/parachain.rs:
```diff
@@ -78,6 +78,67 @@ pub type ValidatorPair = validator_app::Pair;
+	/// May only be called on a state that `ensure_can_swap` has previously returned `Ok` for: if this is
+	/// not the case, the result is undefined. May only return an error if `ensure_can_swap` also returns
+	/// an error.
+	fn on_swap(one: Id, other: Id) -> Result<(), &'static str>;
```
it's also strange to me to call this `on_swap`, when it's rather asking the trait to `do_swap` or `perform_swap` or just plain old `swap` two parachains. We usually use `OnX` traits as event handlers.

gavofyork replied:
it is an event: two paras are being swapped. there may be things that need to be done outside of the entity that is actually managing the swap, so this is a hook to do those things.

rphmeier replied:
typically an event handler is something that is informed when an event has occurred, but cannot determine whether that event is going to occur or not

gavofyork replied:
the `on_swap` (i.e. handler) portion of this trait cannot determine if it's going to happen.

rphmeier replied:
yes, that's the interface as written. However, the result is undefined if `can_swap` has not returned `Ok(())`. That means that whatever carries out the event has three options:
  - Execute `on_swap` for all event handlers without asking about `can_swap`, risking triggering undefined behavior
  - Execute `can_swap` for all event handlers after applying the event, and simply don't trigger `on_swap` for those which cannot swap. This is just strange, since the event handler could just ignore internally, if that's the desired goal.
  - Execute `can_swap` for all event handlers before applying the event, and if any cannot handle it, do not apply the event. That is not a proper event handler, because it alters the control flow of the notifier.

gavofyork replied:
Will rename `SwapAuxilliary`?

gavofyork replied:
done.

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -238,34 +174,12 @@ decl_storage! {
-		/// Did the parachain heads get updated in this block?
-		DidUpdate: bool;
+		/// The ordered list of ParaIds that have a `RelayDispatchQueue` entry.
+		NeedsDispatch: Vec<ParaId>;
```
more documentation here about under what circumstances it is expected not to have a `RelayDispatchQueue` entry would be helpful

gavofyork replied:
from the existing docs:

> Messages ready to be dispatched onto the relay chain.

therefore, this is the ordered list of para ids that have "messages ready to be dispatched onto the relay chain".

not being in here therefore means that the para has no "messages ready to be dispatched onto the relay chain".

is that clear?

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -431,6 +384,11 @@ impl<T: Trait> Module<T> {
+		// we sort them in order to provide a fast lookup to ensure we can avoid duplicates in the
+		// needs_dispatch queue.
+		let mut ordered_needs_dispatch = NeedsDispatch::get();
+		ordered_needs_dispatch.sort_unstable();
```
the documentation on `NeedsDispatch` says that it is sorted - either it is wrong or this line is redundant

gavofyork replied:
they should be sorted, but since there's no way of proving that here, this is a defensive measure to ensure things don't needlessly break if for whatever reason state is corrupted.

gavofyork replied:
yeah it's redundant. removing.

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -452,7 +410,11 @@ impl<T: Trait> Module<T> {
+			Self::queue_upward_messages(
+				id,
+				&head.candidate.upward_messages,
+				&ordered_needs_dispatch[..]
```
```suggestion
				&ordered_needs_dispatch[..],
```

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -471,27 +437,29 @@ impl<T: Trait> Module<T> {
+			let _ = RelayDispatchQueue::append(id, upward_messages);
+			if ordered_needs_dispatch.binary_search(&id).is_err() {
+				// same.
+				let _ = NeedsDispatch::append(&[id]);
```
if the invariant that `NeedsDispatch` should be sorted is to be maintained, you should `let Err(idx) = binary_search(...)` and insert at the index, rather than appending.

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -516,26 +484,32 @@ impl<T: Trait> Module<T> {
+				0
+			} else {
+				(validator_count - 1) / parachain_count
+			};
```
```suggestion
			}
```

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -1213,6 +1241,26 @@ mod tests {
 	}
 
+	fn init_block() {
+		println!("Initializing {}", System::block_number());
```
there are some committed `println!` here and elsewhere

gavofyork replied:
they're in tests and quite handy in case things break for whatever reason.

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Main parachains logic. For now this is just the determination of which validators do what.
```
copypasta from `parachains.rs`

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
+		initial_head_data: Vec<u8>
```
```suggestion
		initial_head_data: Vec<u8>,
```

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
+		initial_head_data: Vec<u8>
```
```suggestion
		initial_head_data: Vec<u8>,
```

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+	}
+}
+
+pub fn swap_ordered_existence<T: PartialOrd + Ord + Copy>(ids: &mut Vec<T>, one: T, other: T) {
```
missing docs

rphmeier replied:
also should take `&[T]` - it doesn't seem that the slice being allocated on the heap or being dynamically sized is a pre-requisite of the interface

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+			#[compact] id: ParaId,
+			code: Vec<u8>,
+			initial_head_data: Vec<u8>,
+			info: ParaInfo
```
```suggestion
			info: ParaInfo,
```

rphmeier replied:
Luke's point about argument ordering still stands

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		}
+
+		/// Deregister a parachain with given id
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000)]
```
this is actually an `O(n)` operation. might be optimizable with a `double_map`

gavofyork replied:
well, insofar as it's a direct dispatch, it's a root call, and there are many other instances of O(N) or worse operation on root dispatches; we assume that the governance can live with the costs if it's dispatching them.

it is, however, used by `deregister_parathread`, which is more problematic. we probably just want to either cancel all queued messages, limit the queue size or require the queue be processed in full before deregistration. i'm not sure which is best, but i'm also not convinced that incoming message queue management (which needs to be addressed even for regular parachains that for whatever reason have sporadic dispatching) is really in the scope of this PR.

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		/// `code` and `initial_head_data` are used to initialize the parathread's state.
+		fn register_parathread(origin,
+			code: Vec<u8>,
+			initial_head_data: Vec<u8>
```
```suggestion
			initial_head_data: Vec<u8>,
```

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		fn select_parathread(origin,
+			#[compact] _id: ParaId,
+			_collator: CollatorId,
+			_head_hash: T::Hash
```
```suggestion
			_head_hash: T::Hash,
```

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+
+				let _ = T::OnSwap::on_swap(id, other);
+			} else {
+				PendingSwap::insert(id, other);
```
there is no corresponding `remove` call for this map (at least insofar as searching `remove` in this file yields nothing). There should be a unit test to check that deregistration cleans up as expected.

rphmeier commented on runtime/src/parachains.rs:
```diff
@@ -274,32 +188,39 @@ decl_module! {
 		/// Provide candidate receipts for parachains, in ascending order by id.
 		#[weight = SimpleDispatchInfo::FixedNormal(1_000_000)]
-		fn set_heads(origin, heads: Vec<AttestedCandidate>) -> Result {
+		pub fn set_heads(origin, heads: Vec<AttestedCandidate>) -> Result {
```
the documentation here says that the heads must be in ascending order by ID. the `on_finalize` function of registrar seems to assume that parathreads are at the end of the `DidUpdate` vec (doesn't appear to be checked by this function).

https://github.com/paritytech/polkadot/pull/341/files#diff-6ceafc4b4dd2600a9e5caebc167c0cecR374

gavofyork replied:
yeah docs needed updating.

gavofyork replied:
perhaps you missed the line 196?

```
active_parachains.sort_by_key(|&(ref id, _)| *id);
```

gavofyork replied:
This should be resolved with my latest commit.

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+				for sched in active {
+					match i {
+						// Scheduled parachain proceeded properly. Move onto next item.
+						Some(para) if para == sched.0 => i = proceeded.next(),
```
This makes a few assumptions:
  1. that `active` is in ascending (well, from this perspective, descending, since we've reversed it) order by para ID. it doesn't seem guaranteed by `on_initialize`.
  2. that `proceeded` has all parathreads at the end (not guaranteed by interface or implementation fo `parachains::set_heads`)

gavofyork replied:
indeed parathread IDs are not guaranteed to be greater than parachain IDs (you're right about the swap mechanism precluding this).

however `Active` is (now) documented as "Ordered as Parachains (ordered by ascending ID) ++ Selected_Parathreads (the last entry is always the item from the retry queue, if one exists)". as such only the first part is sorted. the trick with reversing it is so that we can quickly (O(N)) iterate through the parathreads and figure out whether they were progressed or not.

gavofyork replied:
Resolved in my latest commit.

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+}
+
+impl<T: Trait> Module<T> {
+	pub fn ensure_thread_id(id: ParaId) -> Option<ParaInfo> {
```
```suggestion
        /// Ensures that the given `ParaId` corresponds to a registered parathread, and returns a descriptor if so.
	pub fn ensure_thread_id(id: ParaId) -> Option<ParaInfo> {
```

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+				let selected_threads = &mut upcoming_selected_threads[i];
+				let thread_count = ThreadCount::get() as usize;
+				ensure!(
+					selected_threads.len() < thread_count,
```
since we've already pushed onto `selected_threads` at this point, could it be `selected_threads.len() <= thread_count`?

rphmeier commented on runtime/src/registrar.rs:
```diff
@@ -0,0 +1,1346 @@
+		fn on_initialize() {
+			let next_up = SelectedThreads::mutate(|t| {
+				let r = if t.len() >= QUEUE_SIZE {
+					// Take the first set of parachains in queue
```
```suggestion
					// Take the first set of parathreads in queue
```

nitpick :)

gavofyork commented on runtime/src/parachains.rs:
```diff
@@ -324,36 +247,23 @@ decl_module! {
-		fn on_finalize(_n: T::BlockNumber) {
-			assert!(<Self as Store>::DidUpdate::take(), "Parachain heads must be updated once in the block");
+		fn on_finalize() {
+			assert!(<Self as Store>::DidUpdate::exists(), "Parachain heads must be updated once in the block");
```
can be optimised by replacing with a `::take`, but that would mean refactoring so that `registrar::on_finalize` is always called before `parachains::on_finalize`. i'd prefer to leave such optimisations for later.

shawntabrizi commented:
Some fixes to help this branch compile can be found here:
https://github.com/paritytech/polkadot/pull/470

Although, I ran into some issues and will need to talk to a rusty person.

gavofyork commented:
> ensuring ordering of active_parachains matches expectations (the set_heads extrinsic expects them to be ordered), while on_finalized of registrar expects something else

Addressed in 78d6b53.

> weight of deregistration is off - it is O(n)

It's part of a wider issue that existed before and i'm not sure if here is quite the right place to manage it.

> storage leak of pending swap map (unless I missed something)

Fixed.

> other nits and grumbles from myself and Shawn

Generally addressed.
